### PR TITLE
deno_fetch attempt 2

### PR DESCRIPTION
**Pull request by bartlomieju**
Depends on https://github.com/denoland/deno/pull/7521

### CODE REVIEW

**Review by ry — APPROVED** [review]
LGTM

File: op_crates/fetch/lib.rs [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+
+  // Check scheme before asking for net permission
+  let scheme = url_.scheme();
+  if scheme != "http" && scheme != "https" {
+    return Err(type_error(format!("scheme '{}' not supported", scheme)));
+  }
+
+  {
+    let state_ = state.borrow();
+    let permissions = state_.borrow::<Rc<FP>>();
```
Comment (bartlomieju) [inline op_crates/fetch/lib.rs:114]: had to put `Rc<FP>` here which is bad - it implcitily expects implementor to put `FP` in an Rc... However I couldn't figure out how to code that in the type

File: op_crates/fetch/lib.rs [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+  #[serde(rename_all = "camelCase")]
+  #[serde(default)]
+  struct CreateHttpClientOptions {
+    ca_file: Option<String>,
+  }
+
+  let args: CreateHttpClientOptions = serde_json::from_value(args)?;
+
+  if let Some(ca_file) = args.ca_file.clone() {
+    let permissions = state.borrow::<Rc<FP>>();
```
Comment (bartlomieju) [inline op_crates/fetch/lib.rs:229]: Ditto

File: cli/build.rs [hunk 3]
```diff
@@ -112,6 +117,10 @@ fn main() {
     "cargo:rustc-env=DENO_WEB_LIB_PATH={}",
     deno_web::get_declaration().display()
   );
+  println!(
+    "cargo:rustc-env=DENO_FETCH_LIB_PATH={}",
+    deno_fetch::get_declaration().display()
+  );
```
Comment (ry) [inline cli/build.rs:123]: This method of passing the declaration file is not so elegant. I hope we can find a better solution in the future.

Comment (bartlomieju) [reply cli/build.rs:123]: I agree, if it's build dependencies there's surely a way to include it without using `env!()` macro

**Review by ry — APPROVED** [review]
LGTM

We've been working on untangling fetch for many weeks - ultimately it required a bit of a sledge hammering to get it out. This patch is not the most elegant, but it does the job.

File: cli/tests/unit/blob_test.ts [hunk 4]
```diff
@@ -69,6 +70,7 @@ unitTest(function nativeEndLine(): void {
 
   assertEquals(blob.size, Deno.build.os === "windows" ? 12 : 11);
 });
+*/
```
Comment (ry) [inline cli/tests/unit/blob_test.ts:73]: I remembered you created an issue with all the tests to fix? Can you link me to it?

Comment (bartlomieju) [reply cli/tests/unit/blob_test.ts:73]: https://github.com/denoland/deno/issues/7540

File: op_crates/fetch/lib.rs [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+where
+  FP: FetchPermissions + 'static,
+{
+  #[derive(Deserialize, Default, Debug)]
+  #[serde(rename_all = "camelCase")]
+  #[serde(default)]
+  struct CreateHttpClientOptions {
+    ca_file: Option<String>,
+  }
+
```
Comment (lucacasonato) [inline op_crates/fetch/lib.rs:227]: This is missing a --unstable check.

Comment (ry) [reply op_crates/fetch/lib.rs:227]: There's no unstable check currently
https://github.com/denoland/deno/blob/cead79f5b8ffd376d339b6e0c30e872bfe6820f6/cli/ops/fetch.rs#L174-L191

Comment (lucacasonato) [reply op_crates/fetch/lib.rs:227]: Oh that is bad - there is meant to be one. `Deno.createHttpClient()` is unstable.

**Pull request merged**